lib/server.js

```diff
@@ -0,0 +1,205 @@
+  }
+
+  /**
+   * The main request handler handles all incoming requests.
+   * Integrate with your own server by calling this method and routing all requests to it.
+   * @param {http.IncomingMessage} req
+   * @param {http.ServerResponse} res
+   * @public
+   */
+  async handle(req, res) {
+    if (req.method !== 'GET' && req.method !== 'HEAD') {
+      return this.write(res, 405, 'Method not allowed')
+    }
+
+    // @ts-ignore req.url is fine
+    const paths = url.parse(req.url)?.path?.split('/')
+
+    if (!paths || paths.length < 3) {
+      return this.write(res, 404, 'Malformed request')
+    }
+
+    try {
+      const [, receivedDigest, hex] = paths
+      var decodedUrl = this.verifyHMAC(receivedDigest, hex)
+    } catch (err) {
+      return this.write(res, 403, 'Bad signature')
+    }
+
+    try {
+      var validUrl = await SafeHttpClient.checkUrl(decodedUrl)
+    } catch (err) {
+      // @ts-ignore err does have message
+      return this.write(res, 400, err.message)
+    }
+
+    const controller = new AbortController()
+    const {signal} = controller
+
+    req.on('close', () => {
+      controller.abort()
+    })
+
+    try {
+      // TODO: respect forwarded headers (check if not private IP)
+      const filterRequestHeaders = filterHeaders(REQUEST_HEADERS)
+      const filterResponseHeaders = filterHeaders(RESPONSE_HEADERS)
+      const client = new SafeHttpClient(this.config.maxSize)
+      const {buffer, headers: resHeaders} = await client.safeFetch(validUrl, {
+        // @ts-ignore we can convert req.headers to Headers
+        headers: filterRequestHeaders(new Headers(req.headers)),
+        method: req.method,
+        // We can't blindly follow redirects as the initial checkUrl
+        // might have been safe, but the redirect location might not be.
+        // SafeHttpClient will check the redirect location before following it.
+        redirect: 'manual',
+        signal
+      })
+
+      const headers = {
+        ...DEFAULT_HEADERS,
+        ...filterResponseHeaders(resHeaders),
+        Via: this.config.serverName
+      }
+
+      if (req.method === 'HEAD') {
+        res.writeHead(204, headers)
+        return res.end()
+      }
+      res.writeHead(200, headers)
+      res.write(buffer)
+      return res.end()
+    } catch (err) {
+      if (err instanceof HttpError) {
+        return this.write(res, err.statusCode, err.message)
+      }
+      if (err instanceof Error) {
+        if (err.name === 'AbortError') {
+          return
+        }
+        const msg = err.message || 'Internal server error'
+        return this.write(res, 500, msg)
+      }
+    }
+  }
+
+  /**
+   * @param {string} receivedDigest
+   * @param {string} hex
+   * @private
+   */
+  verifyHMAC(receivedDigest, hex) {
+    // Hex-decode the URL
+    const originalUrlBuffer = hexDecode(hex)
+
+    // Convert the buffer to a string
+    const decodedUrl = originalUrlBuffer.toString('utf-8')
```

Comment:
```suggestion
    // Hex-decode the URL
    const decodedUrl = Buffer.from(hex, 'hex').toString()
```

Simplify, buffers can hex decode. The `utf-8` part is the default also.